Add mask operation on graph and filter graph primitive

amatsukawa opened the pull request:
Implementation for a "project" function and use it in the filter graph primitive.

Closes #39.

rxin commented on graph/src/main/scala/org/apache/spark/graph/Graph.scala:
```diff
@@ -197,6 +197,16 @@ abstract class Graph[VD: ClassManifest, ED: ClassManifest] {
 
+  /**
+   * Take the subgraph of the this graph such that the structure of the result is the same
+   * as the input graph other.
```
I understand you are trying to reference the variable "other", but this sentence is a little bit awkward to read as is. Can we rephrase it?

rxin commented on graph/src/main/scala/org/apache/spark/graph/Graph.scala:
```diff
@@ -197,6 +197,16 @@ abstract class Graph[VD: ClassManifest, ED: ClassManifest] {
+   * @return a graph with vertices and edges that exists in both the current graph and other,
+   * with vertex and edge data from the current graph.
+   */
+  def project[VD2: ClassManifest, ED2: ClassManifest](other: Graph[VD2, ED2]): Graph[VD, ED]
```
If we don't need VD2, ED2, you can do 

```
def project(other: Graph[_, _]): Graph[VD, ED]
```

rxin commented on graph/src/main/scala/org/apache/spark/graph/GraphOps.scala:
```diff
@@ -233,4 +233,37 @@ class GraphOps[VD: ClassManifest, ED: ClassManifest](graph: Graph[VD, ED]) {
+   * }}}
+   *
+   */
+
```
Remove the extra space here.

amatsukawa commented:
re: two type wildcards, scala doesn't seem to like the second one:

```
[error] /Users/amatsukawa/misc/spark-graphx/graph/src/main/scala/org/apache/spark/graph/Graph.scala:208: _ is already defined as type _
[error]   def project[_, _](other: Graph[_, _]): Graph[VD, ED]
[error]                  ^
[error] /Users/amatsukawa/misc/spark-graphx/graph/src/main/scala/org/apache/spark/graph/impl/GraphImpl.scala:258: _ is already defined as type _
[error]   override def project[_, _](other: Graph[_, _]): Graph[VD, ED] = {
```

rxin commented:
Remove the parametrized type 

i.e. 

```
def project(other: Graph[_, _]): Graph[VD, ED]
```

amatsukawa commented:
Ah, sry, misread the comment. Done.

jegonzal commented:
After doing some more in depth profiling analysis we have found some performance bottlenecks in the main graph representation that we are tuning now.  As soon as that stabilizes I would like to look at incorporating this pull request.  Sorry about the delay.

amatsukawa commented:
Sorry I haven't updated this PR in a while, I've had a busy week at work.
I'll try to address Joey's suggestions in the next few nights.

On Sun, Nov 10, 2013 at 11:03 AM, Joey notifications@github.com wrote:

> After doing some more in depth profiling analysis we have found some
> performance bottlenecks in the main graph representation that we are tuning
> now. As soon as that stabilizes I would like to look at incorporating this
> pull request. Sorry about the delay.
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/amplab/graphx/pull/48#issuecomment-28157660
> .

jegonzal commented:
No problem.  Actually, I was suggesting that we have been swamped which is
why we haven't had a chance to discuss the PR. So take your time :-).

amatsukawa commented on graph/src/main/scala/org/apache/spark/graph/impl/GraphImpl.scala:
```diff
@@ -247,6 +247,38 @@ class GraphImpl[VD: ClassManifest, ED: ClassManifest] protected (
+        .map{case (vid, opt) => (vid, opt.get)}
+    )
+
+    val otherImpl = other.asInstanceOf[GraphImpl[VD2, ED2]]
```
Is there a better way of doing this? Tried down casting by pattern matching like so but the compiler didn't like it.

```
other match {
  case otherImpl: GraphImpl => ...
}
```

amatsukawa commented:
Rebased off master. Addressed previous comments by Joey. Implemented an indexed join for vertices, and a local join for edges. 

Currently, projectGraph test is dying with this error. Kryo serializer is set at beginning of GraphSuite. Any idea why this is failing?

```
Job aborted: Task not serializable: java.io.NotSerializableException: org.apache.spark.graph.impl.GraphImpl
org.apache.spark.SparkException: Job aborted: Task not serializable: java.io.NotSerializableException: org.apache.spark.graph.impl.GraphImpl
    at org.apache.spark.scheduler.DAGScheduler$$anonfun$abortStage$1.apply(DAGScheduler.scala:819)
    at org.apache.spark.scheduler.DAGScheduler$$anonfun$abortStage$1.apply(DAGScheduler.scala:817)
    at scala.collection.mutable.ResizableArray$class.foreach(ResizableArray.scala:60)
    at scala.collection.mutable.ArrayBuffer.foreach(ArrayBuffer.scala:47)
    at org.apache.spark.scheduler.DAGScheduler.abortStage(DAGScheduler.scala:817)
    at org.apache.spark.scheduler.DAGScheduler.org$apache$spark$scheduler$DAGScheduler$$submitMissingTasks(DAGScheduler.scala:609)
    at org.apache.spark.scheduler.DAGScheduler$$anonfun$handleTaskCompletion$16.apply(DAGScheduler.scala:726)
    at org.apache.spark.scheduler.DAGScheduler$$anonfun$handleTaskCompletion$16.apply(DAGScheduler.scala:724)
    at scala.collection.mutable.ResizableArray$class.foreach(ResizableArray.scala:60)
    at scala.collection.mutable.ArrayBuffer.foreach(ArrayBuffer.scala:47)
    at org.apache.spark.scheduler.DAGScheduler.handleTaskCompletion(DAGScheduler.scala:724)
    at org.apache.spark.scheduler.DAGScheduler.processEvent(DAGScheduler.scala:429)
    at org.apache.spark.scheduler.DAGScheduler.org$apache$spark$scheduler$DAGScheduler$$run(DAGScheduler.scala:494)
    at org.apache.spark.scheduler.DAGScheduler$$anon$1.run(DAGScheduler.scala:158)
```

ankurdave commented:
That happens because Scala is sometimes overzealous about closure capture. The closures in GraphImpl.mask() need ClassManifests for VD and ED, so they are capturing the surrounding class as well.

The workaround is to move the definition to the GraphImpl object, as has been done for e.g. GraphImpl.mapReduceTriplets.

amatsukawa commented:
Thx @ankurdave made suggested change and test works now.

AmplabJenkins commented:
Can one of the admins verify this patch?

Commit:
add mask operation to Graph

Commit:
add filter operation to GraphOps

amatsukawa commented:
Rebased lots of changes from master. Now I am unable to run my tests, they seem to fail due to this assertion [here](https://github.com/amplab/graphx/blob/master/core/src/main/scala/org/apache/spark/util/collection/OpenHashSet.scala#L173) (is this new?)

```
13/12/04 16:50:09 INFO executor.Executor: Running task ID 4
13/12/04 16:50:09 INFO spark.CacheManager: Partition rdd_11_0 not found, computing it
13/12/04 16:50:09 INFO storage.BlockFetcherIterator$BasicBlockFetcherIterator: Getting 1 non-zero-bytes blocks out of 1 blocks
13/12/04 16:50:09 INFO storage.BlockFetcherIterator$BasicBlockFetcherIterator: Started 0 remote gets in  0 ms
13/12/04 16:50:09 INFO storage.MemoryStore: ensureFreeSpace(784) called with curMem=696, maxMem=85772206
13/12/04 16:50:09 INFO storage.MemoryStore: Block rdd_11_0 stored as values to memory (estimated size 784.0 B, free 81.8 MB)
13/12/04 16:50:09 INFO storage.BlockManagerMasterActor$BlockManagerInfo: Added rdd_11_0 in memory on 172.25.135.137:53995 (size: 784.0 B, free: 81.8 MB)
13/12/04 16:50:09 INFO storage.BlockManagerMaster: Updated info of block rdd_11_0
13/12/04 16:50:09 INFO storage.BlockFetcherIterator$BasicBlockFetcherIterator: Getting 1 non-zero-bytes blocks out of 1 blocks
13/12/04 16:50:09 INFO storage.BlockFetcherIterator$BasicBlockFetcherIterator: Started 0 remote gets in  1 ms
13/12/04 16:50:09 INFO storage.BlockManager: Found block rdd_11_0 locally
13/12/04 16:50:09 INFO storage.BlockFetcherIterator$BasicBlockFetcherIterator: Getting 1 non-zero-bytes blocks out of 1 blocks
13/12/04 16:50:09 INFO storage.BlockFetcherIterator$BasicBlockFetcherIterator: Started 0 remote gets in  0 ms
13/12/04 16:50:09 INFO local.LocalTaskSetManager: Loss was due to java.lang.AssertionError
java.lang.AssertionError: assertion failed
    at scala.Predef$.assert(Predef.scala:146)
    at org.apache.spark.util.collection.OpenHashSet.getValueSafe(OpenHashSet.scala:173)
    at org.apache.spark.util.collection.OpenHashSet.getValueSafe$mcJ$sp(OpenHashSet.scala:172)
    at org.apache.spark.graph.impl.VertexPartition$$anonfun$map$1.apply$mcVI$sp(VertexPartition.scala:29)
    at org.apache.spark.graph.impl.VertexPartition$$anonfun$map$1.apply(VertexPartition.scala:28)
    at org.apache.spark.graph.impl.VertexPartition$$anonfun$map$1.apply(VertexPartition.scala:28)
    at scala.collection.Iterator$class.foreach(Iterator.scala:772)
    at org.apache.spark.util.collection.BitSet$$anon$1.foreach(BitSet.scala:122)
    at org.apache.spark.graph.impl.VertexPartition.map(VertexPartition.scala:28)
    at org.apache.spark.graph.impl.GraphImpl$$anonfun$mapVertices$1.apply(GraphImpl.scala:181)
    at org.apache.spark.graph.impl.GraphImpl$$anonfun$mapVertices$1.apply(GraphImpl.scala:181)
    at scala.collection.Iterator$$anon$19.next(Iterator.scala:401)
    at org.apache.spark.graph.VertexSetRDD$$anonfun$3.apply(VertexSetRDD.scala:160)
    at org.apache.spark.graph.VertexSetRDD$$anonfun$3.apply(VertexSetRDD.scala:158)
    at org.apache.spark.rdd.ZippedPartitionsRDD2.compute(ZippedPartitionsRDD.scala:89)
    at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:237)
    at org.apache.spark.rdd.RDD.iterator(RDD.scala:226)
    at org.apache.spark.rdd.MapPartitionsRDD.compute(MapPartitionsRDD.scala:36)
    at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:237)
    at org.apache.spark.rdd.RDD.iterator(RDD.scala:226)
    at org.apache.spark.rdd.FlatMappedRDD.compute(FlatMappedRDD.scala:32)
    at org.apache.spark.graph.VertexSetRDD.compute(VertexSetRDD.scala:134)
    at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:237)
    at org.apache.spark.rdd.RDD.iterator(RDD.scala:226)
    at org.apache.spark.rdd.MappedRDD.compute(MappedRDD.scala:29)
    at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:237)
    at org.apache.spark.rdd.RDD.iterator(RDD.scala:226)
    at org.apache.spark.rdd.MapPartitionsRDD.compute(MapPartitionsRDD.scala:36)
    at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:237)
    at org.apache.spark.rdd.RDD.iterator(RDD.scala:226)
    at org.apache.spark.scheduler.ShuffleMapTask.runTask(ShuffleMapTask.scala:159)
    at org.apache.spark.scheduler.ShuffleMapTask.runTask(ShuffleMapTask.scala:100)
    at org.apache.spark.scheduler.Task.run(Task.scala:53)
    at org.apache.spark.executor.Executor$TaskRunner$$anonfun$run$1.apply$mcV$sp(Executor.scala:215)
    at org.apache.spark.deploy.SparkHadoopUtil.runAsUser(SparkHadoopUtil.scala:50)
    at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:182)
    at java.util.concurrent.ThreadPoolExecutor$Worker.runTask(ThreadPoolExecutor.java:895)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:918)
    at java.lang.Thread.run(Thread.java:695)
```

Any idea how to resolve this?

ankurdave commented:
#81 broke the tests; I'd like to look at this tonight.

amatsukawa commented:
Hmm, but only my 2 tests seem to be breaking.

ankurdave commented:
Ah, you're right. The errors were just compile errors in GraphSuite and SerializerSuite (which I just fixed in #83). @dcrankshaw mentioned a similar error as what you're seeing, though, so it's a bug in GraphX. I'll look into it.

dcrankshaw commented:
I just pulled in #81 and am seeing reproducible errors when trying to deserialize Longs. I'm chasing it down now.

ankurdave commented:
Could you push the rebased version of this PR (to this branch or a different branch) so I can take a look at the error?

amatsukawa commented:
@dcrankshaw @ankurdave have you had a chance to take a look at the error yet? Trying to fix this now...

ankurdave commented:
I think the error was fixed by #88, but correct me if I'm wrong.

amatsukawa commented:
It looks the code for GraphImpl changed somewhat (eTable is gone now?) so I'll have to update this again. I'll let you know. Let's try to merge this in ASAP it's getting very difficult to keep this branch up to date with so many changes going on in GraphImpl.

amatsukawa commented:
I get huge merge conflicts every time and basically have to re-make the change manually every time.

ankurdave commented:
Agreed, sorry for the delay. Many more changes are coming soon, and I'll try and merge this before they get merged.

BTW, eTable is merely renamed to edges.

amatsukawa commented:
@ankurdave I hear from @jegonzal that you guys are going to take a look at merging in the PRs? Do you want to try to make the changes then? Might be much easier for you instead of me trying to cargo-cult from the subgraph method, which is what I do each tim haha.

Looks like VertexSetRDD is also gone, although I suspect it's been renamed to VertexRDD? Were there type changes as well? Looks like the changes aren't simple renaming, because RDD transformation functions don't type-match anymore.

ankurdave commented:
Sure, I can handle updating this PR. In that case I'll merge the other changes in first to avoid having to update this code twice.

ankurdave commented on graph/src/main/scala/org/apache/spark/graph/GraphOps.scala:
```diff
@@ -244,4 +244,35 @@ class GraphOps[VD: ClassManifest, ED: ClassManifest](graph: Graph[VD, ED]) {
+   *
+   */
+  def filter[VD2: ClassManifest, ED2: ClassManifest](
+      preprocess: Graph[VD, ED] => Graph[VD2, ED2],
```
Why not just pass in the preprocessed graph directly?

ankurdave replied:
Oh, I guess the point of the function is to do the three operations (preprocess, subgraph, and mask) all in one. Considering that you have to pass in all the arguments to those functions as well, though, I wonder whether this would really make user code more understandable.

amatsukawa replied:
The thinking was to abstract away the intermediate graph that we actually
filter on, because this is a fairly common usage pattern (you want to
filter the original graph, but based on some property you need to compute).

I agree it's not a major change, though.

On Mon, Dec 16, 2013 at 11:29 PM, Ankur Dave notifications@github.comwrote:

> In graph/src/main/scala/org/apache/spark/graph/GraphOps.scala:
> 
> > -   \* changing the vertex and edge values in your program. For example, we could remove the vertices
> > -   \* in a graph with 0 outdegree
> > -   *
> > -   \* {{{
> > -   \* graph.filter(
> > -   \*   graph => {
> > -   \*     val degrees: VertexSetRDD[Int] = graph.outDegrees
> > -   \*     graph.outerJoinVertices(degrees) {(vid, data, deg) => deg.getOrElse(0)}
> > -   \*   },
> > -   \*   vpred = (vid: Vid, deg:Int) => deg > 0
> > -   \* )
> > -   \* }}}
> > -   *
> > -   */
> > -  def filter[VD2: ClassManifest, ED2: ClassManifest](
> > -      preprocess: Graph[VD, ED] => Graph[VD2, ED2],
> 
> Oh, I guess the point of the function is to do the three operations
> (preprocess, subgraph, and mask) all in one. Considering that you have to
> pass in all the arguments to those functions as well, though, I wonder
> whether this would make user code more understandable.
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/amplab/graphx/pull/48/files#r8393978
> .

ankurdave committed:
Reimplement Graph.mask using innerJoin

ankurdave committed:
Merge remote-tracking branch 'upstream/master' into add_project_to_graph

Conflicts:
	graph/src/main/scala/org/apache/spark/graph/Graph.scala
	graph/src/main/scala/org/apache/spark/graph/impl/GraphImpl.scala

ankurdave commented:
I merged the current master into this PR and updated it to work. @amatsukawa, could you merge my PR on your repo?

amatsukawa commented:
Hey Ankur, Thx, I'll update my repo. So I can delete this PR now right?

On Wed, Dec 18, 2013 at 1:03 PM, Ankur Dave notifications@github.comwrote:

> I merged the current master into this PR and updated it to work.
> @amatsukawa https://github.com/amatsukawa, could you merge my PR on
> your repo?
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/amplab/graphx/pull/48#issuecomment-30880465
> .

ankurdave commented:
Don't delete this PR -- if you merge amatsukawa/graphx#1, it should update this PR, which I can then merge.

amatsukawa commented:
Oohh, ok I see. Will do thx.

On Wed, Dec 18, 2013 at 2:43 PM, Ankur Dave notifications@github.comwrote:

> Don't delete this PR -- if you merge amatsukawa/graphx#1https://github.com/amatsukawa/graphx/pull/1,
> it should update this PR, which I can then merge.
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/amplab/graphx/pull/48#issuecomment-30888717
> .

amatsukawa committed:
Merge pull request #1 from ankurdave/add_project_to_graph

Merge current master and reimplement Graph.mask using innerJoin

ankurdave commented:
Thanks. I'll merge this once the Jenkins tests finish.

amatsukawa commented:
Great, thanks Ankur :)
